clang/lib/Sema/SemaTypeTraits.cpp

```diff
@@ -2399,6 +2400,66 @@ static void DiagnoseNonConstructibleReason(
   SemaRef.Diag(D->getLocation(), diag::note_defined_here) << D;
 }
 
+static void DiagnoseNonDestructibleReason(Sema &SemaRef, SourceLocation Loc,
+                                          QualType T) {
+
+  QualType CoreT = T.getCanonicalType();
+  if (const ArrayType *AT = SemaRef.Context.getAsArrayType(CoreT))
+    CoreT = AT->getElementType();
+
+  SemaRef.Diag(Loc, diag::note_unsatisfied_trait)
+      << CoreT << diag::TraitName::Destructible;
+
+  if (CoreT->isFunctionType()) {
```

Comment:
There is so much repetition on these next handful of branches.  It looks like some sort of work shoudl be done to only do the SemaRef.Diag 1x and store the TraitNotSatisifedReason to then diagnose it.